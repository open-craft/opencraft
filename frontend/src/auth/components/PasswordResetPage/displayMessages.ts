const messages = {
  password: {
    defaultMessage: 'Password*',
    description: ''
  },
  passwordHelp: {
    defaultMessage:
      'Password must have 8 characters, 1 number, 1 capital letter and 1 special character.',
    description: ''
  },
  passwordConfirm: {
    defaultMessage: 'Confirm password*',
    description: ''
  },
  passwordConfirmHelp: {
    defaultMessage: 'Confirm the password in the field above.',
    description: ''
  },
  reset: {
    defaultMessage: 'Reset',
    description: ''
  },
  passwordsDoNotMatch: {
    defaultMessage: 'The entered passwords do not match.',
    description: ''
  },
  validatingToken: {
    defaultMessage:
      "We're currently validating the provided password reset token. Please wait.",
    description: ''
  },
  invalidToken: {
    defaultMessage: 'The token you provided is invalid or expired.',
    description: ''
  }
};

export default messages;
