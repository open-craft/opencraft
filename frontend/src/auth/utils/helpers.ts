import { V2Api } from 'global/api';

export const checkAuthAndRefreshToken = async () => {
  const refreshToken = localStorage.getItem('token_refresh');
  let response = null;

  if (refreshToken) {
    try {
      response = await V2Api.authRefreshCreate({
        data: { refresh: refreshToken }
      });
    } catch (e) {
      return false;
    }
  }

  if (response) {
    localStorage.setItem('token_access', response.access);
    return true;
  }

  return false;
};
