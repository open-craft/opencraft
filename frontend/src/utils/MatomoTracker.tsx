import React, { ReactNode } from 'react';
import {
  createInstance,
  MatomoProvider,
  useMatomo
} from '@datapunt/matomo-tracker-react';
import {
  MATOMO_ALIAS_DOMAIN,
  MATOMO_BASE_URL,
  MATOMO_MY_DOMAIN,
  MATOMO_SITE_ID
} from '../global/constants';

interface MatomoTrackerProps {
  children: ReactNode;
}

interface MatomoAdditionalConfiguration {
  [propName: string]: any;
}

interface MatomoConfiguration {
  urlBase: string;
  siteId: number;
  configurations?: MatomoAdditionalConfiguration;
}

interface MatomoUserIdTrackerProps {
  userId: string;
}

/*
 * This wrapper component initializes the Matomo tracking by wrapping the children
 * elements, if the tracking is enabled by providing a valid value for the 'MATOMO_BASE_URL'
 * constant. It allows tracking page views, events and the user id, when available.
 *
 * The 'MATOMO_ALIAS_DOMAIN' constant allows the cross-domain linking of another site with
 * this site so that they are treated as the same site in Matomo.
 */
export const MatomoTracker: React.FC<MatomoTrackerProps> = (
  props: MatomoTrackerProps
) => {
  let element = props.children;
  if (MATOMO_BASE_URL) {
    const matomoConfiguration: MatomoConfiguration = {
      urlBase: MATOMO_BASE_URL,
      siteId: parseInt(MATOMO_SITE_ID, 10)
    };
    if (MATOMO_ALIAS_DOMAIN) {
      const myDomain = `*.${MATOMO_MY_DOMAIN}`;
      const aliasDomain = `*.${MATOMO_ALIAS_DOMAIN}`;
      matomoConfiguration.configurations = {
        enableCrossDomainLinking: true,
        setDomains: [myDomain, aliasDomain]
      };
    }
    element = (
      <MatomoProvider value={createInstance(matomoConfiguration)}>
        {props.children}
      </MatomoProvider>
    );
  }
  return <>{element}</>;
};

/**
 * This is a component passes the provided user id to Matomo if Matomo tracking is enabled.
 * The 'pushInstruction' function exposed by the 'useMatomo()' hook is used to do this.
 *
 * This component is needed as all the components in the registration process are classes and hence
 * the 'yseMatomo' hook cannot be directly used in those components.
 */
export const MatomoUserIdTracker: React.FC<MatomoUserIdTrackerProps> = (
  props: MatomoUserIdTrackerProps
) => {
  const { pushInstruction } = useMatomo();
  if (MATOMO_BASE_URL && props.userId) {
    pushInstruction('setUserId', props.userId);
  }
  return null;
};
