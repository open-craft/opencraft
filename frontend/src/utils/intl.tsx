import React from "react";
import { FormattedMessage, MessageDescriptor } from "react-intl";

type MessageMap = { [k: string]: Omit<MessageDescriptor, 'id'> };

interface Props<T, K extends keyof T> {
    id: K;
    messages: T;
    values?: any;
}


export const WrappedMessage = (props: Props<MessageMap, string>) => {
    const {messages, id, ...values} = props;
    return (
        <FormattedMessage  {...messages[id]} id={id} {...values}/>
    );
};
