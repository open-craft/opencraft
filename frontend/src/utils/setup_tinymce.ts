import 'tinymce/tinymce';
import 'tinymce/themes/silver';
import 'tinymce/skins/ui/oxide/skin.min.css';
import 'tinymce/skins/ui/oxide/content.min.css';
import 'tinymce/plugins/advlist';
import 'tinymce/plugins/autolink';
import 'tinymce/plugins/lists';
import 'tinymce/plugins/link';
import 'tinymce/plugins/image';
import 'tinymce/plugins/charmap';
import 'tinymce/plugins/print';
import 'tinymce/plugins/preview';
import 'tinymce/plugins/anchor';
import 'tinymce/plugins/searchreplace';
import 'tinymce/plugins/visualblocks';
import 'tinymce/plugins/code';
import 'tinymce/plugins/fullscreen';
import 'tinymce/plugins/insertdatetime';
import 'tinymce/plugins/media';
import 'tinymce/plugins/table';
import 'tinymce/plugins/paste';
import 'tinymce/plugins/help';
import 'tinymce/plugins/wordcount';
