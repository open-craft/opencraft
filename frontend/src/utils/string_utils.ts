export const toCamelCase = (s: string) => {
  return s.replace(/([-_][a-z])/gi, ($1) => {
    return $1.toUpperCase().replace('-', '').replace('_', '');
  });
};

export const sanitizeErrorFeedback = (input: {
  [key: string]: Array<string>;
}) => {
  // Loop at each error message and join them.
  // Also convert keys from snake_case to camelCase
  const newObject: any = {};

  Object.keys(input).forEach((key) => {
    newObject[toCamelCase(key)] = input[key].join();
  });

  return newObject;
};

export const capitalizeFirstLetter = (str: string) => {
  return str[0].toUpperCase() + str.slice(1);
};
