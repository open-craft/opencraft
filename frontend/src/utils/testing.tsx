import { createMemoryHistory } from 'history';
import * as React from 'react';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import * as renderer from 'react-test-renderer';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import { createRootReducer } from '../global/reducers';

export const setupComponentForTesting = (
  reactContent: JSX.Element,
  storeContents = {}
) => {
  const middleware = applyMiddleware(thunk);

  const store = createStore(
    createRootReducer(createMemoryHistory()),
    storeContents,
    middleware
  );

  return renderer.create(
    <IntlProvider textComponent={React.Fragment} locale="en">
      <Provider store={store}>
        <MemoryRouter>{reactContent}</MemoryRouter>
      </Provider>
    </IntlProvider>
  );
};
