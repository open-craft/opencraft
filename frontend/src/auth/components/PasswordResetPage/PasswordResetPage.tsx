import * as React from 'react';
import { RootState } from 'global/state';
import { connect } from 'react-redux';
import { WrappedMessage } from 'utils/intl';
import { Alert, Button, Spinner, Form } from 'react-bootstrap';
import { ContentPage, TextInputField } from 'ui/components';
import {
  clearErrorMessage,
  performPasswordReset,
  performPasswordResetTokenValidation
} from 'auth/actions';
import messages from './displayMessages';
import './styles.scss';

interface ActionProps {
  clearErrorMessage: Function;
  performPasswordResetTokenValidation: Function;
  performPasswordReset: Function;
}

interface State {
  [key: string]: string;
  password: string;
  passwordConfirm: string;
}

interface Props extends ActionProps {
  // Retrieve token from URL
  match: {
    params: {
      token: string;
    };
  };
  loading: boolean;
  error: string;
  succeeded: boolean;
}

@connect<{}, ActionProps, {}, Props, RootState>(
  (state: RootState) => state.loginState,
  {
    clearErrorMessage,
    performPasswordResetTokenValidation,
    performPasswordReset
  }
)
export class PasswordResetPage extends React.PureComponent<Props, State> {
  constructor(props: Props, state: State) {
    super(props);

    this.state = {
      password: '',
      passwordConfirm: ''
    };

    this.props.performPasswordResetTokenValidation({
      token: this.props.match.params.token
    });
  }

  private validatePasswords = () => {
    return (
      this.state.password && this.state.password === this.state.passwordConfirm
    );
  };

  private onKeyPress = (event: any) => {
    if (event.key === 'Enter' && this.validatePasswords()) {
      event.preventDefault();
      this.submit();
    }
  };

  private submit = () => {
    this.props.performPasswordReset({
      token: this.props.match.params.token,
      password: this.state.password
    });
  };

  private onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const field = e.target.name;
    const { value } = e.target;

    // Clear error message when user changes a field
    this.setState({
      [field]: value
    });

    this.props.clearErrorMessage();
  };

  public render() {
    return (
      <ContentPage title="Reset your password">
        {!this.props.succeeded &&
          (!this.props.error ? (
            <div>
              <Spinner animation="border" size="sm" className="spinner" />
              <WrappedMessage messages={messages} id="validatingToken" />
            </div>
          ) : (
            <WrappedMessage messages={messages} id="invalidToken" />
          ))}

        {this.props.succeeded && (
          <Form
            className="login" // FIXME: make SCSS more generic to properly reuse this style.
            onKeyPress={(event: any) => {
              this.onKeyPress(event);
            }}
          >
            <TextInputField
              fieldName="password"
              value={this.state.password}
              onChange={this.onChange}
              type="password"
              messages={messages}
            />
            <TextInputField
              fieldName="passwordConfirm"
              value={this.state.passwordConfirm}
              onChange={this.onChange}
              type="password"
              messages={messages}
            />

            {this.props.error && (
              <Alert variant="danger">{this.props.error}</Alert>
            )}

            {this.state.password !== this.state.passwordConfirm && (
              <Alert variant="danger">
                <WrappedMessage messages={messages} id="passwordsDoNotMatch" />
              </Alert>
            )}

            <Button
              className="pull-left loading"
              variant="primary"
              size="lg"
              disabled={this.props.loading || !this.validatePasswords()}
              onClick={() => {
                this.submit();
              }}
            >
              {this.props.loading && (
                <Spinner animation="border" size="sm" className="spinner" />
              )}
              <WrappedMessage messages={messages} id="reset" />
            </Button>
          </Form>
        )}
      </ContentPage>
    );
  }
}
