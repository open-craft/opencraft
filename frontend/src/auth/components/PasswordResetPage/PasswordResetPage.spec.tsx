import React from 'react';
import { setupComponentForTesting } from "utils/testing";
import { PasswordResetPage } from './PasswordResetPage';

it('renders without crashing', () => {
    const mockTokenProps = {
    params: {
      token: "test_token"
    }
  }
    const tree = setupComponentForTesting(
      <PasswordResetPage
        match={mockTokenProps}
        loading={false}
        succeeded={false}
        error={''}
        clearErrorMessage={() => {}}
        performPasswordReset={() => {}}
        performPasswordResetTokenValidation={() => {}}
      />).toJSON();
    expect(tree).toMatchSnapshot();
});
