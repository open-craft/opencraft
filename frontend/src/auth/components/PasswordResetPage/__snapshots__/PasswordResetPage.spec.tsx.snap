// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="content-page"
>
  <div
    className="title-container"
  >
    <h1>
      Reset your password
    </h1>
  </div>
  <div
    className="content-page-container"
  >
    <div
      className="content-page-content row"
    >
      <div>
        <div
          className="spinner spinner-border spinner-border-sm"
        />
        We're currently validating the provided password reset token. Please wait.
      </div>
    </div>
  </div>
</div>
`;
